src/client.rs

```diff
@@ -827,7 +885,13 @@ impl HttpClient {
             }
         }
 
-        // Set custom request headers.
+        // Check if user has setup headers in defaults already
+        if let Some(extension) = self.defaults.get::<DefaultHeaderMap>() {
+            for (name, value) in extension.0.iter() {
+                parts.headers.insert(name, value.clone());
+            }
+        }
```

Comment:
If we just have a normal `default_headers: http::HeaderMap<http::HeaderValue>` field, then we can apply them by calling `self.default_headers.set_opt(&mut easy)?`.

Something I'm not sure about is how this interacts with duplicate headers. Should headers present in the request _override_ default headers, or should they be included _in addition_ to the default headers?

Reply:
>If we just have a normal default_headers: http::HeaderMap<http::HeaderValue> field, then we can apply them by calling self.default_headers.set_opt(&mut easy)?

I actually started implementing this with `set_opt` itself but then realized that `.set_opt(&mut easy)?` will actually mutate the internal curl client. Then lets say later we call `parts.headers.set_opt(&mut easy)` the headers get overridden by this new call and default headers will get completely discarded.

> Something I'm not sure about is how this interacts with duplicate headers. Should headers present in the request override default headers, or should they be included in addition to the default headers?

I believe this is a similar situation where someone calls `default_header` twice. In that case we consider this an insert or append situation?
```
HttpClientBuilder::new()
            .default_header("myheader", "value1")
            .default_header("myheader", "value2") ## < ---- Append or Insert ?
            .build();
```

Reply:
> I actually started implementing this with `set_opt` itself but then realized that `.set_opt(&mut easy)?` will actually mutate the internal curl client. Then lets say later we call `parts.headers.set_opt(&mut easy)` the headers get overridden by this new call and default headers will get completely discarded.

Good point!

> I believe this is a similar situation where someone calls `default_header` twice. In that case we consider this an insert or append situation?

I've been thinking about this, and also looking at other clients to see if there is a precedent for this behavior. I think it would make sense for `default_header` to act as an `append`, but when _applying_ the default headers, I would expect a default header value to only be used if not present in the request object.

Reply:
Here are some examples of what behavior would seem intuitive to me, let me know what you think.

```rust
HttpClientBuilder::new()
    .default_header("foo", "bar")
    .build()?
    .send(Request::get("")
        .build()?)?
// headers -- foo: bar
```

```rust
HttpClientBuilder::new()
    .default_header("foo", "bar")
    .build()?
    .send(Request::get("")
        .header("foo", "baz")
        .build()?)?
// headers -- foo: baz
```

```rust
HttpClientBuilder::new()
    .default_header("foo", "bar")
    .default_header("foo", "baz")
    .build()?
    .send(Request::get("")
        .build()?)?
// headers -- foo: bar, foo: baz
```

```rust
HttpClientBuilder::new()
    .default_header("foo", "bar")
    .default_header("foo", "baz")
    .build()?
    .send(Request::get("")
        .header("foo", "x")
        .header("foo", "y")
        .build()?)?
// headers -- foo: x, foo: y
```

Reply:
This definitely looks clean to me. `Request` definitely gets the priority here.

Reply:
[`HeaderMap::keys`](https://docs.rs/http/0.2.1/http/header/struct.HeaderMap.html#method.keys) might also help with the implementation here.

Reply:
I have updated the code but kind of confused about this one failing test.  Used the `Matcher::Any` for `user-agent` matching in mockito, but it weirdly fails still.

Reply:
I _think_ this is a bug in Mockito when multiple headers with the same name are encountered. I am looking into it now.

Reply:
Looks like this is a misunderstanding of Mockito on my part, I created an issue there: https://github.com/lipanski/mockito/issues/117

Based on some manual tests I think this is working correctly, so I'd just comment out the tests for now until Mockito supports this kind of assertion.